Assets/Scripts/Game/Guilds/GuildManager.cs

```diff
@@ -258,7 +259,7 @@ public void RestoreMembershipData(Dictionary<int, GuildMembership_v1> data)
             ClearMembershipData();
             if (data != null)
             {
-                foreach (FactionFile.GuildGroups guildGroup in data.Keys)
+                foreach (FactionFile.GuildGroups guildGroup in data.Keys.OfType<FactionFile.GuildGroups>())
```

Comment:
Is this really neccessary?

Reply:
`OfType` is useful to filter by type old collections like `ArrayList`, which are now deprecated with the introduction of generics.

I believe `OfType` is not doing anything here because the cast is always succesful. If you want to check that an int value is defined with an enum, you need to use something like `.Where(x => Enum.IsDefined(typeof(FactionFile.GuildGroups), x))`, but this comes with a big performance cost.

Reply:
I'll undo this change. Sorry, I'd meant to bring this one up when submitting the PR but forgot to.

Here is the web site for the warning that led to me changing that:
https://rules.sonarsource.com/csharp/RSPEC-3217

They call it "critical" and it sounded like a worthwhile change/guideline to keep to, but unless we have a general consensus for it I'll ignore this one.

Reply:
Foreach loop makes a cast due to [historical reasons](https://ericlippert.com/2013/07/22/why-does-a-foreach-loop-silently-insert-an-explicit-conversion/). It's a bad idea to make use of it because it can hide a possibly bad design, so this is a good point to rise.

The cast done here is from int to enum, which is always valid. The issue is that it's easy to forget to handle unknown values in an appropriate way. @ajrb might want to check this is the case and maybe see if values can be directly received and stored as enum instead of getting int values.

Reply:
Thanks for the insight @TheLacus. I'll leave this one to @ajrb if he would like to pursue it further and decide whether to make any changes.

Reply:
I'll take a look at some point, thanks for the info.